# TSE Data Analysis

In [5]:
library(dplyr)
library(ggplot2)


Attaching package: ‘dplyr’


The following objects are masked from ‘package:stats’:

    filter, lag


The following objects are masked from ‘package:base’:

    intersect, setdiff, setequal, union




In [17]:
prs <- read.csv('../data/final.csv', stringsAsFactors=TRUE)
nrow(prs)
head(prs)
summary(prs)

[1] 3786

X owner   name   number cursor           repoUrl                          
1 1 abel533 Mapper 245    Y3Vyc29yOjIwMA== https://github.com/abel533/Mapper
2 2 abel533 Mapper 577    Y3Vyc29yOjIwMA== https://github.com/abel533/Mapper
3 3 abel533 Mapper 647    Y3Vyc29yOjIwMA== https://github.com/abel533/Mapper
4 4 abel533 Mapper 741    Y3Vyc29yOjIwMA== https://github.com/abel533/Mapper
5 5 abel533 Mapper 743    Y3Vyc29yOjIwMA== https://github.com/abel533/Mapper
6 6 abel533 Mapper 793    Y3Vyc29yOjIwMA== https://github.com/abel533/Mapper
  clients prUrl                                     
1 53441   https://github.com/abel533/Mapper/pull/245
2 53441   https://github.com/abel533/Mapper/pull/577
3 53441   https://github.com/abel533/Mapper/pull/647
4 53441   https://github.com/abel533/Mapper/pull/741
5 53441   https://github.com/abel533/Mapper/pull/743
6 53441   https://github.com/abel533/Mapper/pull/793
  title                                     state ⋯ base   breakingChanges
1 添加别名                                  OPEN  ⋯ master 0              
2 Update SqlHelper.java                     OPEN  ⋯ master 0              
3 增加指定字段强制更新(包含null)方法        OPEN  ⋯ master 0              
4 Feat:增加批量update功能                   OPEN  ⋯ master 0              
5 修复 AggregationProvider 的问题           OPEN  ⋯ master 0              
6 feat:新增根据实体中的属性进行查询指定属性 OPEN  ⋯ master 0              
  brokenClients brokenUses changedFiles checkedClients
1 0             0          7            0             
2 0             0          1            0             
3 0             0          2            0             
4 0             0          6            0             
5 0             0          5            0             
6 0             0          2            0             
  errors                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  
1 null[package] failed (1): [ERROR] COMPILATION ERROR : [ERROR] Source option 6 is no longer supported. Use 7 or later.[ERROR] Target option 6 is no longer supported. Use 7 or later.[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.1:compile (default-compile) on project mapper-core: Compilation failure: Compilation failure: [ERROR] Source option 6 is no longer supported. Use 7 or later.[ERROR] Target option 6 is no longer supported. Use 7 or later.[ERROR] -> [Help 1][ERROR] [ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.[ERROR] Re-run Maven using the -X switch to enable full debug logging.[ERROR] [ERROR] For more information about the errors and possible solutions, please read the following articles:[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException                                                                                                                             
2 null[package] failed (1): [ERROR] COMPILATION ERROR : [ERROR] Source option 6 is no longer supported. Use 7 or later.[ERROR] Target option 6 is no longer supported. Use 7 or later.[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-pl

       X                   owner           name          number     
 Min.   :   1.0   apache      :1594   trino  : 390   Min.   :   66  
 1st Qu.: 947.2   trinodb     : 390   alluxio: 232   1st Qu.: 3599  
 Median :1893.5   Alluxio     : 232   hudi   : 177   Median : 8574  
 Mean   :1893.5   OpenAPITools:  93   flink  : 173   Mean   : 9827  
 3rd Qu.:2839.8   alibaba     :  80   pulsar : 159   3rd Qu.:14849  
 Max.   :3786.0   keycloak    :  76   hadoop :  93   Max.   :32312  
                  (Other)     :1321   (Other):2562                  
              cursor                                   repoUrl    
 Y3Vyc29yOjIxNA==: 343   https://github.com/trinodb/trino  : 390  
 Y3Vyc29yOjI0MA==: 214   https://github.com/Alluxio/alluxio: 232  
 Y3Vyc29yOjU1    : 173   https://github.com/apache/hudi    : 177  
 Y3Vyc29yOjM4OA==: 143   https://github.com/apache/flink   : 173  
 Y3Vyc29yOjExMg==: 125   https://github.com/apache/pulsar  : 159  
 Y3Vyc29yOjgz    :  95   https://github.com/ap

## Popular active Java libraries on GitHub, as of 2022-11-13
  - Repositories written in Java
  - With a pom.xml file at the root
  - With >= 5000 stars
  - With >= 100 clients (all packages included)
  - Last push activity >= 2022-10-13 (1 month)
  - PR created >= 2022-10-13 (1 month)
  - PR merged >= 2022-09-13 (2 months)
  - !archived, !disabled, !fork, !mirror, !locked

In [21]:
libs <- distinct(prs[, c('owner', 'name')])
nrow(libs)
head(libs)
summary(libs)

[1] 230

owner    name                   
1 abel533  Mapper                 
2 Activiti Activiti               
3 adobe    aem-core-wcm-components
4 adobe    S3Mock                 
5 airlift  airlift                
6 alibaba  arthas

          owner                          name    
 apache      : 53   druid                  :  2  
 alibaba     : 10   accumulo               :  1  
 spring-cloud: 10   activemq               :  1  
 google      :  5   activemq-artemis       :  1  
 bytedeco    :  3   Activiti               :  1  
 eclipse     :  3   aem-core-wcm-components:  1  
 (Other)     :146   (Other)                :223  

## Interesting pull requests
  - Opened, unmerged, and unlocked
  - Affects Java files
  - Updated >= 2022-10-13 (1 month)

In [22]:
head(prs)
summary(prs)

X owner   name   number cursor           repoUrl                          
1 1 abel533 Mapper 245    Y3Vyc29yOjIwMA== https://github.com/abel533/Mapper
2 2 abel533 Mapper 577    Y3Vyc29yOjIwMA== https://github.com/abel533/Mapper
3 3 abel533 Mapper 647    Y3Vyc29yOjIwMA== https://github.com/abel533/Mapper
4 4 abel533 Mapper 741    Y3Vyc29yOjIwMA== https://github.com/abel533/Mapper
5 5 abel533 Mapper 743    Y3Vyc29yOjIwMA== https://github.com/abel533/Mapper
6 6 abel533 Mapper 793    Y3Vyc29yOjIwMA== https://github.com/abel533/Mapper
  clients prUrl                                     
1 53441   https://github.com/abel533/Mapper/pull/245
2 53441   https://github.com/abel533/Mapper/pull/577
3 53441   https://github.com/abel533/Mapper/pull/647
4 53441   https://github.com/abel533/Mapper/pull/741
5 53441   https://github.com/abel533/Mapper/pull/743
6 53441   https://github.com/abel533/Mapper/pull/793
  title                                     state ⋯ base   breakingChanges
1 添加别名                                  OPEN  ⋯ master 0              
2 Update SqlHelper.java                     OPEN  ⋯ master 0              
3 增加指定字段强制更新(包含null)方法        OPEN  ⋯ master 0              
4 Feat:增加批量update功能                   OPEN  ⋯ master 0              
5 修复 AggregationProvider 的问题           OPEN  ⋯ master 0              
6 feat:新增根据实体中的属性进行查询指定属性 OPEN  ⋯ master 0              
  brokenClients brokenUses changedFiles checkedClients
1 0             0          7            0             
2 0             0          1            0             
3 0             0          2            0             
4 0             0          6            0             
5 0             0          5            0             
6 0             0          2            0             
  errors                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  
1 null[package] failed (1): [ERROR] COMPILATION ERROR : [ERROR] Source option 6 is no longer supported. Use 7 or later.[ERROR] Target option 6 is no longer supported. Use 7 or later.[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.1:compile (default-compile) on project mapper-core: Compilation failure: Compilation failure: [ERROR] Source option 6 is no longer supported. Use 7 or later.[ERROR] Target option 6 is no longer supported. Use 7 or later.[ERROR] -> [Help 1][ERROR] [ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.[ERROR] Re-run Maven using the -X switch to enable full debug logging.[ERROR] [ERROR] For more information about the errors and possible solutions, please read the following articles:[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException                                                                                                                             
2 null[package] failed (1): [ERROR] COMPILATION ERROR : [ERROR] Source option 6 is no longer supported. Use 7 or later.[ERROR] Target option 6 is no longer supported. Use 7 or later.[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-pl

       X                   owner           name          number     
 Min.   :   1.0   apache      :1594   trino  : 390   Min.   :   66  
 1st Qu.: 947.2   trinodb     : 390   alluxio: 232   1st Qu.: 3599  
 Median :1893.5   Alluxio     : 232   hudi   : 177   Median : 8574  
 Mean   :1893.5   OpenAPITools:  93   flink  : 173   Mean   : 9827  
 3rd Qu.:2839.8   alibaba     :  80   pulsar : 159   3rd Qu.:14849  
 Max.   :3786.0   keycloak    :  76   hadoop :  93   Max.   :32312  
                  (Other)     :1321   (Other):2562                  
              cursor                                   repoUrl    
 Y3Vyc29yOjIxNA==: 343   https://github.com/trinodb/trino  : 390  
 Y3Vyc29yOjI0MA==: 214   https://github.com/Alluxio/alluxio: 232  
 Y3Vyc29yOjU1    : 173   https://github.com/apache/hudi    : 177  
 Y3Vyc29yOjM4OA==: 143   https://github.com/apache/flink   : 173  
 Y3Vyc29yOjExMg==: 125   https://github.com/apache/pulsar  : 159  
 Y3Vyc29yOjgz    :  95   https://github.com/ap

## Results

In [26]:
breaking <- subset(all, breakingChanges > 0)
nrow(all)
nrow(breaking)
impacting <- subset(breaking, brokenClients > 0)
nrow(impacting)
nrow(results)

[1] 3786

[1] 465

[1] 83

ERROR: Error in nrow(results): object 'results' not found


Overall, 465 out of 3786 PRs are breaking (12.28%), and 42 have an impact on clients (2.19% of all, 17.85% of breaking ones).

## BreakBot reports
  - First, we fork the original repository in our own organization `breakbot-playground`
  - Then, we copy the PRs of interest in the fork
  - BreakBot runs automatically on these PRs to produce its reports
  - The `.github/breakbot.yml` configuration we use is as follows (100 clients per impacted package at most, with a minimum of 5 stars)
  
```
clients:
  top: 100
  stars: 5
```
